src/main/java/com/openlattice/datastore/search/controllers/SearchController.java

```diff
@@ -555,75 +529,109 @@ public SearchResult executeFQNPropertyTypeSearch( @RequestBody FQNSearchTerm sea
+    @Timed
+    public Map<UUID, Map<UUID, Map<UUID, Set<NeighborEntityIds>>>> executeFilteredEntityNeighborIdsSearch(
+            @RequestBody EntityNeighborsFilterBulk filter ) {
+        Set<Principal> principals = Principals.getCurrentPrincipals();
+
+        Map<UUID, Map<UUID, Map<UUID, Set<NeighborEntityIds>>>> result = Maps.newHashMap();
+        var entitySets = getEntitySetsForRead( filter.getEntityKeyIds().keySet(), principals );
+        if ( entitySets != null ) {
+            var groupedEntitySets = entitySets.entrySet().stream()
+                    .collect( Collectors.groupingBy( esEntry -> esEntry.getValue().isLinking() ) );
+            if ( !groupedEntitySets.get( false ).isEmpty() ) {
+                var regularEntitySetIds = groupedEntitySets.get( false ).stream()
+                        .map( Map.Entry::getKey ).collect( Collectors.toSet() );
+                result = searchGraphService.executeEntityNeighborIdsSearch(
+                        new EntityNeighborsFilterBulk(
+                                filter.getEntityKeyIds().entrySet().stream()
+                                        .filter( entry -> regularEntitySetIds.contains( entry.getKey() ) )
+                                        .collect( Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue ) ),
+                                filter.getSrcEntitySetIds(),
+                                filter.getDstEntitySetIds(),
+                                filter.getAssociationEntitySetIds()
+                        ),
+                        principals );
+            } else if ( !groupedEntitySets.get( true ).isEmpty() ) {
+                var linkedEntitySetIds = groupedEntitySets.get( true ).stream()
+                        .map( Map.Entry::getKey ).collect( Collectors.toSet() );
+                result.putAll(
+                        searchGraphService.executeLinkingEntityNeighborIdsSearch(
+                                groupedEntitySets.get( true ).stream()
+                                        .collect( Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue ) ),
+                                new EntityNeighborsFilterBulk(
+                                        filter.getEntityKeyIds().entrySet().stream()
+                                                .filter( entry -> linkedEntitySetIds.contains( entry.getKey() ) )
+                                                .collect( Collectors.toMap( Map.Entry::getKey, Map.Entry::getValue ) ),
+                                        filter.getSrcEntitySetIds(),
+                                        filter.getDstEntitySetIds(),
+                                        filter.getAssociationEntitySetIds()
+                                ),
+                                principals ) );
             }
         }
 
 
         /* audit */
 
-        SetMultimap<UUID, UUID> neighborsByEntitySet = HashMultimap.create();
+        Map<UUID, Set<UUID>> neighborsByEntitySet = Maps.newHashMap();
```

Comment:
another map sizing, `result.values.size() * associationMap.size() * association.size`

Reply:
yeah, if we want to consider all, we also need to take neighbors into account: `result.size() * associationMap.size() * association.size * neighbors.size()`